src/test/RequestQueue.test.ts

```diff
@@ -218,12 +218,79 @@ describe("test RequestQueue", () => {
         { key: "b", requestAction: work },
       ];
       const start = Date.now();
-      const promises = rq.addRequests(requests, delayMs);
+      const promises = rq.addRequests(requests, false, delayMs);
       rq.addRequest("b", work); // requesting this again should remove the delay
       await Promise.allSettled(promises);
       expect(Date.now() - start).to.be.lessThan(delayMs);
       expect(count).to.equal(2);
     });
+
+    it("queues all regular requests before any low-priority requests", async () => {
+      let regular = 0;
+      let lowPriority = 0;
+      const regularWork = async () => {
+        await sleep(5);
+        regular++;
+      };
+      const lowPriorityWork = async () => {
+        await sleep(5);
+        lowPriority++;
+      };
+
+      const rq = new RequestQueue(1);
+      const prom1 = rq.addRequest("a", regularWork);
+      const prom2 = rq.addRequest("b", lowPriorityWork, true);
+      const prom3 = rq.addRequest("c", regularWork);
+
+      await prom1;
+      expect(regular).to.equal(1);
+      expect(lowPriority).to.equal(0);
+
+      await prom3;
+      expect(regular).to.equal(2);
+      expect(lowPriority).to.equal(0);
+
+      await prom2;
+      expect(lowPriority).to.equal(1);
+    });
+
+    it("maintains a separate and lower concurrent task limit for low-priority requests", async () => {
+      let count = 0;
+      const work = async () => {
+        await sleep(5);
+        count++;
+      };
+
+      const rq = new RequestQueue(2, 1);
+      const prom1 = rq.addRequest("a", work, true);
+      const prom2 = rq.addRequest("b", work, true);
+      expect(rq.requestRunning("a")).to.be.true;
+      expect(rq.requestRunning("b")).to.be.false;
+      rq.addRequest("c", work);
+      expect(rq.requestRunning("c")).to.be.true;
+      await prom1;
+      await prom2;
+      expect(count).to.equal(3);
+    });
+
+    it("can promote a low-priority task to high priority", async () => {
+      let count = 0;
+      const work = async () => {
+        await sleep(5);
+        count++;
+      };
+
+      const rq = new RequestQueue(2, 1);
+      const prom1 = rq.addRequest("a", work);
+      const prom2 = rq.addRequest("b", work, true);
+      expect(rq.requestRunning("a")).to.be.true;
+      expect(rq.requestRunning("b")).to.be.false;
+      rq.addRequest("b", work);
+      expect(rq.requestRunning("b")).to.be.true;
+      await prom1;
+      await prom2;
+      expect(count).to.equal(2);
+    });
```

Comment:
I know the current behavior for RequestQueue requires all high priority tasks to be complete before moving onto the low priority tasks... Would this ever cause problematic stalling?

Like, say a high priority task is stuck due to a network timeout or something. This would prevent all low priority tasks for being executed, even if there is capacity available to handle them.

Reply:
Okay, correction. Took a closer look at RequestQueue, and it allows low priority requests to be started once the total number of active requests dips below the max low priority count and if the high priority queue is empty. Question answered 👍